htlcswitch/link.go

```diff
@@ -3053,6 +3065,25 @@ func (l *channelLink) processRemoteAdds(fwdPkg *channeldb.FwdPkg,
 
 		fwdInfo := pld.ForwardingInfo()
 
+		// If we are in a flush state we need to cancel back all of the
+		// net new HTLCs rather than forwarding them. This is the first
+		// opportunity we have to bounce invalid HTLC adds without
+		// doing a force-close.
+		if l.IsFlushing() {
```

Comment:
IMO this doesn't need to be caught here, it can be caught earlier in `handleUpstreamMsg` when the add is first sent across. It should also only apply if we've received the peer's `Shutdown`

Reply:
Additionally, this should also be caught in `handleDownstreamMsg` or in `handleDownstreamUpdateAdd`

Reply:
I talked to @Roasbeef about this. Earlier versions of this did exactly what you suggest but the issue is a question of how you'd respond in the case when they issued the add. From a code cleanliness perspective I would love to do this in handleUpstreamMsg but the issue is it will result in more aggressive behavior than is necessary.

Reply:
I don't think it's safe to fail HTLCs here. If we've previously forwarded the HTLC it may exist on the outgoing link's commitment tx. If we then startup and call `resolveFwdPkgs`, it will sequentially call `processRemoteAdds` for each non-full forwarding package. If we then call `Flush` at some point during this, we'll cancel one of these HTLCs on the incoming side. Since the outgoing side may still have the HTLC, we'd lose the value of an HTLC

Reply:
My understanding from my conversation with @Roasbeef yesterday was that this is the function responsible for actually locking in upstream HTLCs and then deciding to forward them after they have been locked in.

My understanding of my change here is that if we were in this flushing state when these irrevocable commitments were made to net-new htlc's, then we haven't forwarded them yet.

What I am getting from your comment here is that if we are mid-way through this process of dealing with the remote adds, then we shouldn't be able to put things *into* a flushing state, which I agree with.

That said, I'd like to understand what you mean by "if we've previously forwarded the HTLC..." because as I understand things that shouldn't be possible.

Reply:
On link startup, `resolveFwdPkgs` is called which will reforward adds, settles, fails to ensure that these messages reach the intended link (this link will be different from the one doing the reforwarding). This is done in `processRemoteSettleFails` & in `processRemoteAdds`. We do this because it ensures that if our node crashed or shutdown before these messages reached the other link, we're able to recover on startup. So my point above was that if the add reached the outgoing link and then we end up canceling here due to flushing, we're in a bad state. One way to get around this might be to not cancel if the state is `FwdStateProcessed` (see several lines below), but I'd have to see if there is any funny business with exit hops

Reply:
> So my point above was that if the add reached the outgoing link and then we end up canceling here due to flushing, we're in a bad state.

If we've reached this point after an internal retransmission, then my understanding is that we haven't wrote anything in the forwarding packages for that entry so it's safe to cancel back. Otherwise, the calls to `sendHTLCError` above would introduce the same issue (cancel back HTLC when we have something locked in the outgoing link).

Reply:
The forwarding package is only written to once the _response_ to the add is signed for on the incoming link. So it's possible that an outgoing htlc exists already. The calls to `sendHTLCError` above and below should be reproducible such that if it this is a retransmission and it fails, any earlier invocation should have also failed (otherwise we have a bug).

This isn't the case with the flush logic because an earlier `processRemoteAdds` may have forwarded the HTLC and now if we've come back up and are flushing, we could hit this error while the HTLC is outgoing